text/2020-01-30-key-spaces.md

```diff
@@ -0,0 +1,197 @@
+
+The problem with running multiple applications is conflicting, unsecured key spaces.
+* Avoiding accidental conflicting use of a key space in TiKV is currently ad-hoc and unreliable
+* Securing data in TiKV against privilege escalation. If one TiKV client is hacked, effectively all are hacked.
+
+Once key spaces are in place, all applications and TiKV itself should understand what data belongs to whom and how to avoid key space conflict.
+An auth implementation can then secure the key spaces to provide security against privilege escalation.
+
+## Detailed design
+
+#### Scope
+
+This RFC will not discuss authentication.
+Authentication is necessary to provide any meaningful security for key spaces.
+
+This RFC will not discuss resource isolation.
+Key spaces will be an important tool for resource isolation to leverage.
+
+This RFC will not propose "sub spaces", which is a key space within a key space.
+
+
+#### Client view
+
+When a client first connects to the TiKV system, it will register itself to get a key space prefix.
+The client can register for additional key spaces, but normally one application will have one key space.
+
+All key space operations take place in the metadata component (PD) which stores a registry of key spaces.
+
+The TiKV component does not know about key spaces. All requests to the storage component (TiKV) will include the key space prefix.
+Please see [this separate proposal that helps with sending key prefixes to TiKV](https://github.com/tikv/rfcs/pull/56).
+
+
+#### Metadata operations
+
+POST /keyspaces
+
+Create a key space
+
+requires:
+
+* key space name
+* application name
+
+optional:
+
+* description
+* prefix
+
+system generated:
+
+* prefix
+
+This API will return the key space object. The key space name is the id of the keyspace and cannot be changed.
+The usage of an application name as an identity may end up being a de-normalization of data in a future iteration when true identity with authentication is added.
+
+Normally the request does not include a prefix and allows the system to generate it.
+The request may include a prefix: this is designed to support backwards compatibility. The prefix should be encoded as the system does, as explained in Key Prefix specification.
+
+
+PUT /keyspaces/{id}
+
+Update a key space
+
+requires id and description. Only the description field can be updated.
+
+Addtional GET APIs will also return keyspace objects
+
+* Lookup by ID, key space name
+* List all key spaces
+
+DELETE /keyspaces/{id}
+
+requires id. Deletion is background task. This DELETE API sets the time of the API call in a field `deleted_at`.
+When deletion is completed, it will set a field `delete_completed_at` and the prefix may be re-used in a new key space.
+
+In a token-based auth system (proposed in another RFC) deletion will need to wait until client tokens expire before it can begin deletion.
+
+
+
+#### Key Prefix specification
```

Comment:
I suggest to use id instead of bytes prefix. TiKV doesn't have to store the KV with given prefix, it can totally store them in separate CFs. Not using byte prefix gives more flexibility for implementations and also easy to get backward compatible. Client can use the default key space when non is specified.

Reply:
Column families would lead to increased overhead when a key space is very small? But this is more of an implementation detail that the specification should try to accomodate.

Reply:
For efficiency we may want tiny key spaces to be stored together but to split larger key spaces into column families or RocksDB instances. Currently we don't store each region in its own column family.

TiKV has no knowledge of key spaces in this proposal. I think that we can extend the PD placement rules (or create something similar) to be able to specify for different key ranges storage properties such as whether to use column families or a separate RocksDB instance. These storage properties would need to be sent to TiKV. TiKV would need to route keys to the correct storage based on the key prefix.

I think prefix matching is equivalent to using an id for a prefix. If key space ids are used and stored in the same column family then they must be translated into a prefix.

Reply:
Any id would also be a valid binary prefix. The difference between the two is really whether there is a separate field for the prefix/id or whether it is just prepended. I tried to have a separate field for the prefix (sending via a header) in https://github.com/tikv/rfcs/pull/56, but I then noted:

> There are a lot of places where prefixes could now be used. But there are also already a lot of places that assume a full key (with the 'z' first byte prefix) is already being given.

I am also worried about forgetting to send the prefix.

Reply:
My preference is to use a binary prefix so that we don't require anything new of TiKV and that clients also are not forced to do anything differently. Once auth is enforced clients need to authenticate but otherwise don't have to do anything differently. And any programs that need full key space access (PD, full backup) don't have to do anything differently. We can still take the approach of #56 to allow for sending prefixes separately and variants of that are possible where we use a separate GRPC API field for a key space prefix.